src/imitation/algorithms/bc.py

```diff
@@ -4,178 +4,261 @@
+import tqdm
 from stable_baselines3.common import policies, utils, vec_env
 
 from imitation.algorithms import base as algo_base
-from imitation.algorithms.base import TransitionMapping
 from imitation.data import rollout, types
 from imitation.policies import base as policy_base
-from imitation.util import logger
+from imitation.util import logger as imit_logger
 
 
-def reconstruct_policy(
-    policy_path: str,
-    device: Union[th.device, str] = "auto",
-) -> policies.ActorCriticPolicy:
-    """Reconstruct a saved policy.
+@dataclasses.dataclass(frozen=True)
+class BatchIteratorWithEpochEndCallback:
+    """Loops through batches from a batch loader and calls a callback after every epoch.
 
-    Args:
-        policy_path: path where `.save_policy()` has been run.
-        device: device on which to load the policy.
-
-    Returns:
-        policy: policy with reloaded weights.
+    Will throw an exception when an epoch contains no batches.
     """
-    policy = th.load(policy_path, map_location=utils.get_device(device))
-    assert isinstance(policy, policies.ActorCriticPolicy)
-    return policy
 
+    batch_loader: Iterable[algo_base.TransitionMapping]
+    n_epochs: Optional[int]
+    n_batches: Optional[int]
+    on_epoch_end: Optional[Callable[[int], None]]
 
-class ConstantLRSchedule:
-    """A callable that returns a constant learning rate."""
+    def __post_init__(self):
+        epochs_and_batches_specified = (
+            self.n_epochs is not None and self.n_batches is not None
+        )
+        neither_epochs_nor_batches_specified = (
+            self.n_epochs is None and self.n_batches is None
+        )
+        if epochs_and_batches_specified or neither_epochs_nor_batches_specified:
+            raise ValueError(
+                "Must provide exactly one of `n_epochs` and `n_batches` arguments.",
+            )
 
-    def __init__(self, lr: float = 1e-3):
-        """Builds ConstantLRSchedule.
+    def __iter__(self) -> Iterator[algo_base.TransitionMapping]:
+        def batch_iterator():
+
+            # Note: the islice here ensures we do not exceed self.n_epochs
+            for epoch_num in itertools.islice(itertools.count(), self.n_epochs):
+                num_batches_in_epoch = 0
+                for num_batches_in_epoch, batch in enumerate(self.batch_loader):
+                    yield batch
+
+                if num_batches_in_epoch == 0:
+                    raise AssertionError(
+                        f"Data loader returned no data after "
+                        f"{num_batches_in_epoch} batches, during epoch "
+                        f"{epoch_num} -- did it reset correctly?",
+                    )
+                if self.on_epoch_end is not None:
+                    self.on_epoch_end(epoch_num)
+
+        # Note: the islice here ensures we do not exceed self.n_batches
+        return itertools.islice(batch_iterator(), self.n_batches)
+
+
+@dataclasses.dataclass(frozen=True)
+class BehaviorCloningLoss:
```

Comment:
+1 on having this be a dataclass not just some unstructured dict. But is it really the loss? In line 127 we have `loss = neglogp + ent_loss + l2_loss`, so the other values are either intermediate values (`entropy`, `l2_norm` unweighted versions of `ent_loss` and `l2_loss`) or are purely information (`prob_true_act` I don't see used anywhere in computation of final `loss`).

Perhaps rename it to `BehaviorCloningStats` or similar?

Reply:
You are right. This is not really the loss and I am also not happy with the name. `BehaviorCloningStats` is not wrong but too generic IMO. What do you think about `BCPerformanceStats`, `BCLossStats` (but then again its more than just loss)? Maybe you can come up with a better word than `Performance` here?

Reply:
`BCPerformanceStats` seems better than `BehaviorCloningLoss`. Another option is `BCTrainingMetrics`?

Reply:
I like `BCTrainingMetrics` I will go with that!